numpy/core/src/private/ufunc_override.h

```diff
@@ -153,6 +154,64 @@ normalize_at_args(PyUFuncObject *ufunc, PyObject *args,
 }
 
 /*
+ * Check args for a object with a `__numpy_ufunc__` attribute.
+ */
+static int
+check_tuple_for_numpy_ufunc(PyObject *args, int *noa,
+                            int *i, PyObject *with_override[NPY_MAXARGS],
+                            int with_override_pos[NPY_MAXARGS]) {
+    int og_i = *i; /* original i */
+    int og_noa = *noa; /* original number of overriding args */
+    int nargs = PyTuple_GET_SIZE(args);
+    PyObject *obj;
+
+    for (; *i < (og_i + nargs); ++*i) {
+
+        obj = PyTuple_GET_ITEM(args, *i);
+
+        /* short circuit optimization for common cases. */
+        if (PyArray_CheckExact(obj) || PyArray_IsScalar(obj, Generic) ||
+            _is_basic_python_type(obj)) {
+            continue;
+        }
+
+        if (PyObject_HasAttrString(obj, "__numpy_ufunc__")) {
+            with_override[*noa] = obj;
+            with_override_pos[*noa] = *i;
+            ++*noa;
+        }
+    }
+    if ((noa - og_noa) > 0) {
+        return 1;
+    }
+    return 0;
+}
+
+static int
+check_kwds_for_numpy_ufunc(PyObject *kwds, int *noa, int *i,
+                           PyObject **with_override,
+                           int with_override_pos[NPY_MAXARGS]) {
+    PyObject *obj;
+    if ((kwds)&& (PyDict_CheckExact(kwds))) {
+        obj = PyDict_GetItemString(kwds, "out");
+        if (obj != NULL) {
+            if (PyObject_HasAttrString(obj, "__numpy_ufunc__")) {
```

Comment:
This may be a microoptimization, but similarly to what you do in `check_tuple_for_numpy_ufunc`, you could skip searching the object's dictionary for the string if it is `None` or a strict `ndarray` or scalar.